### PR TITLE
#503 | EIP-1967 Proxy contracts interface

### LINKED ISSUE #503: implementation contract interface for proxy contracts
## Overview
For proxy contracts that point to an implementation contract there is no contract interface. In these cases, we want to fetch the implementation contract address and either redirect or display source interface in an additional tab. 
## Links
example: https://www.teloscan.io/address/0xb99C43d3bce4c8Ad9B95a4A178B04a7391b2a6EB#contract
bsc explorer example with read/write as proxy tabs: https://bscscan.com/address/0x8ac76a51cc950d9822d68b83fe1ad97b32cd580d#readContract

### PR DESCRIPTION
**Pull request by ezra-sg**
# Fixes #503 

## Description
This PR enables users to interface with proxy contracts in the Contract tab of the address page.

In order to test this functionality, I have deployed some contracts on mainnet:
base contract: [`0x86A2fA5eA80A831A935ED2fAF3B3ad2F11a04Ff6`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x86A2fA5eA80A831A935ED2fAF3B3ad2F11a04Ff6#contract)
proxy contract with no methods: [`0x0cB032Af8Ff931647eD838856af731a6FBc68F7D`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x0cB032Af8Ff931647eD838856af731a6FBc68F7D#contract)
proxy contract with methods: [`0xbbd1170a77F9eA0458872AF740204e7C8f9391f1`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xbbd1170a77F9eA0458872AF740204e7C8f9391f1#contract)

base contract with overloaded methods: [`0xA42dCbe6EDdACd25BEb54B5cC8CBf273c14B5898`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xA42dCbe6EDdACd25BEb54B5cC8CBf273c14B5898#contract)
unverified proxy contract: [`0x8A44f083b1BA3112C776EF7194F572977c684c37`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x8A44f083b1BA3112C776EF7194F572977c684c37#contract)
proxy for contract with overloads: [`0xAE21A25fd0aA5e7a366ba42EfC65685c0acDdC53`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xAE21A25fd0aA5e7a366ba42EfC65685c0acDdC53#contract)
proxy with updateable implementation: [`0x359CC3AA5134379D296c369DE7fc417C6C4e3e5B`](https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x359CC3AA5134379D296c369DE7fc417C6C4e3e5B#contract)

**Update**: I misunderstood how proxy and implementation contracts interact when i first created this PR. I have updated the testing notes. When you call a method on an implementation contract through a proxy (from the `Read as Proxy` or `Write as Proxy` tab), the data is stored **on the proxy**. So when you call `increment` **on the proxy**, it does **not** update the `count` on the implementation base contract. the implementation contract maintains its own state which is not shared with the proxy contract. So when you call `count()` on the proxy contract using the `Read as Proxy` tab, you are not reading the implementation contract's state! You are actually reading state stored in the proxy contract in a special implementation slot. The proxy contract also "has its own state", which is why `incrementProxyCount` and `getProxyCount` (in the `Read` and `Write` tabs) do not affect the state called `count`.

## Test scenarios
- go to https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/
- log in with metamask
- go to the address page for a proxy contract whose implementation contract is _unverified_, such as https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xb99C43d3bce4c8Ad9B95a4A178B04a7391b2a6EB#contract
    - you should see 2 new tabs, 'Read as Proxy' and 'Write as Proxy'
- go to the read or write tab for the contract given above (note that for this contract, there are no executable methods)
    - you should see the message "No executable functions found in the contract ABI"
- go to the 'Read as Proxy' tab
    - you should see a message saying that "This contract is an EIP-1967 Transparent Proxy for the contract 0x1701D9B92BAf8C5d01F97CA97b328Cc4db945E93"
    - you should see another message saying "The contract for which this is a proxy is not verified, so its ABI is not available. To interact with the implementation contract, visit the implementation contract page"
    - the links should go to the expected page
- go to 'Write as Proxy' tab
    - you should see the same thing as 'Read as Proxy' tab
- go to the address page of a proxy contract whose implementation contract _is verified_, e.g. https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xbbd1170a77F9eA0458872AF740204e7C8f9391f1#contract
- go to the 'Read' tab
    - you should see some executable methods, which are exposed by the contract
- execute `getProxyCount` and take note of the number
    - this number is tracked by the contract itself, NOT on the implementation contract, i.e. this proxy contract has its own methods in addition to being a proxy
- go to the 'Write' tab
- execute `incrementProxyCount`
- go back to the 'Read' tab
- execute `getProxyCount`
    - the result should be 1 greater than the last time you called it
- go to 'Read as Proxy' tab
- execute `count`, which is a count variable stored by the _implementation_ contract
- in a new tab, go to the implementation contract page, https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0x86A2fA5eA80A831A935ED2fAF3B3ad2F11a04Ff6#contract
- go to the 'Read' tab
- execute `count` and take note of the result
- go back to the proxy contract page
- go to the 'Write as Proxy' tab
- execute `increment`
- go back to 'Read as Proxy' tab
- execute `count`
    - the count should have incremented
- go back to your tab with the implementation contract
- execute `count`
    - the count should **not have changed**
- on the implementation contract page, go to the 'Write' tab
- execute `increment`
- go to the 'Write' tab
- execute `count` and take note of the result
- go back to your tab with the proxy contract
- go to the 'Read' tab
- execute `count`
    - the count should **not have changed** from the last time you looked at the proxy count

## Checklist:

-   [ ] I have performed a self-review of my own code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have cleaned up the code in the areas my change touches
-   [ ] My changes generate no new warnings
-   [ ] Any dependent changes have been merged and published in downstream modules
-   [ ] I have checked my code and corrected any misspellings
-   [ ] I have removed any unnecessary console messages
-   [ ] I have included all english text to the translation file and/or created a new issue with the required translations for the currently supported languages
-   [ ] I have tested for mobile functionality and responsiveness
-   [ ] I have added appropriate test coverage

### CODE REVIEW

**Comment by pmjanus** [comment]
image
had this error when I sent a transaction in metamask with less gas, the error message was:
Error Details
{ "error": "{\"code\":-32603,\"message\":\"Internal JSON-RPC error.\"}" }

wonder where the error message displayed comes from since that function I used has 0 parameters ( incrementProxyCount on https://deploy-preview-520--dev-mainnet-teloscan.netlify.app/address/0xe9B3E9F3b9f54EDa2455254262241695B808A5f3#contract )

**Comment by pmjanus** [comment]
Will this support non-EIP-1967 ?
Diamon Proxies for example.

Can we test for Proxy Contract Upgrades, Transparent and UUPS ?

**Comment by pmjanus** [comment]
Another thing we want to test is function overloading, does the ui handle it?

File: src/i18n/es-es/index.js [hunk 1]
```diff
@@ -329,6 +330,12 @@ export default {
             read_functions: 'Funciones de lectura',
             write_functions: 'Funciones de escritura',
             unverified_contract: 'Contrato no verificado',
+            read_as_proxy: 'Leer como proxy',
+            write_as_proxy: 'Escribir como proxy',
+            contract_is_proxy: 'Este contrato es un proxy transparente EIP-1967 para el contrato',
+            proxy_for_unverified_contract_part_1: 'El contrato para el cual este es un proxy no está verificado, por lo que su ABI no está disponible. Para interactuar con el contrato de implementación, visite el',
```
Comment (Viterbo) [inline src/i18n/es-es/index.js:336]:
```suggestion
            proxy_for_unverified_contract_part_1: 'El contrato para el cual este es un proxy no está verificado, por lo que su ABI no está disponible. Para interactuar con el contrato de implementación, visite la',
```

**Review by Viterbo — APPROVED** [review]
Nice work! :+1:

**Comment by Viterbo** [comment]
Currently, they differ in the `count` value:

image

image

**Comment by Viterbo** [comment]
Everything works as expected !!
It would be nice to see what happens when you change the implementation. I couldn't do it because I'm not the owner.
image

**Review by Viterbo — APPROVED** [review]

File: src/components/ContractTab/ContractTab.vue [hunk 2]
```diff
@@ -30,55 +49,114 @@ export default {
...
-            return this.source === false && this.write === true;
+    },
+    methods: {
+        async checkEIP1967Implementation() {
+            if (!this.contractAddress) {
+                return '';
+            }
+
+            const web3 = new Web3(process.env.NETWORK_EVM_RPC);
+
```
Comment (pmjanus) [inline src/components/ContractTab/ContractTab.vue:68]: https://ethereum-blockchain-developer.com/110-upgrade-smart-contracts/09-eip-1967/

**Review by pmjanus — APPROVED** [review]

**Pull request merged**